postgresql/helpers.go

```diff
@@ -92,6 +92,45 @@ func revokeRoleMembership(db QueryAble, role, member string) error {
 	return nil
 }
 
+// withRolesGranted temporary grants, if needed, the roles specified to connected user
+// (i.e.: the admin configure in the provider) and revoke them as soon as the
+// callback func has finished.
+func withRolesGranted(txn *sql.Tx, roles []string, fn func() error) error {
+	// Get current user
+	var currentUser string
+	err := txn.QueryRow("SELECT CURRENT_USER").Scan(&currentUser)
+	switch {
+	case err == sql.ErrNoRows:
+		return fmt.Errorf("SELECT CURRENT_USER returns now row, this is quite disturbing")
+	case err != nil:
+		return errwrap.Wrapf("error while looking for the current user: {{err}}", err)
+	}
+
+	var granted []string
+	for _, role := range roles {
+		roleGranted, err := grantRoleMembership(txn, role, currentUser)
+		if err != nil {
+			return err
+		}
```

Comment:
If you succeed in granting some roles but fail at some point, you will return an error without revoking the roles you already granted. You may want to wrap the revocation of a list of roles in a function that gets called if an error is raised here.

Reply:
Ho didn't see this comment before answering to the other :facepalm: .

Everything is executed in a transaction, the caller will not be able to commit a failed transaction.
(that's why I accept only `*sql.Tx` and not a `Queryable` interface like other helpers.